content/polyhedra/functions.ts

```diff
@@ -21,36 +21,175 @@ import './components/folding';
+    // or 91 degrees, or a non-right angle with a label of 90.
+    //
+    if (adjustableAngleNames.indexOf(midPointName) >= 0) {
+      if (Math.abs(deg - 90) > 1) {
+        adjustThisPoint = midPointName;
+      }
+      // else {
+      //   console.log(midPointName, exact, exact.isRight, deg, Math.abs(deg - 90));
+      // }
+    }
+
+    sum += deg;
+    roundedSum += rounded;
+    geometry[midPointName] = {
+      exact,
+      rounded,
+    };
+  };
+
+  const delta = expectedSum - roundedSum;
+  if (delta !== 0 ) {
+    // console.log('Mismatch', adjustableAngleNames, adjustThisPoint, geometry[adjustThisPoint].rounded, roundedSum, expectedSum, delta, roundedSum, sum);
+    geometry[adjustThisPoint].rounded += delta;
+  }
+
+  geometry.sum = sum;
+  geometry.roundedSum = roundedSum;
+  geometry.overlap = (Math.round(sum) > expectedSum);
+
+  return geometry;
+}
+
 
+// computeChangedAngles()
+// This function would not be necessary if I knew a better way to obtain
+// the currently dragged point(s) in a GeoPad. So what this function does is to
+// figure out which points have changed position and to return an array of the
+// relevant angles that have changed, which includes the adjacent angles to a given
+// dragged point.
+//
+// So if one point is being dragged, then three angles are adjustable.
+// If a line segment (two points) were to be dragged, then four angles would be affected.
+//
+// This function compares oldPoints with the latest values in the model to determine
+// what (if anything) changed. It completes by updating oldPoints with the latest model
+// values.
+//
+
+function computeChangedAngles(model: Observable, pointNames: string[], oldPoints: object[]): string[] {
+  const adjustableAngleNames = [];
+  const numPoints = pointNames.length;
+
+  for (let index = 0; index < numPoints; ++index) {
+    const prevPointName = pointNames[index === 0 ? numPoints - 1 : index - 1];
+    const midPointName = pointNames[index];
+    const nextPointName = pointNames[index === numPoints - 1 ? 0 : index + 1];
+
+    const newPoint: GeoPoint = model[midPointName];
+    const newPointCopy = {x: newPoint.x, y: newPoint.y};
+    const oldPoint = oldPoints[midPointName] || newPointCopy;
+    if (oldPoint.x !== newPointCopy.x || oldPoint.y !== newPointCopy.y) {
+      adjustableAngleNames.push(prevPointName);
+      adjustableAngleNames.push(midPointName);
+      adjustableAngleNames.push(nextPointName);
+    }
+    oldPoints[midPointName] = newPointCopy;
+  };
+
+  return adjustableAngleNames;
+}
+
+
+export function angles($step: Step): void {
+  const model = $step.model;
+  const leftPolygon = {
+    modelKey: 'geometryLeft',
+    pointNames: ['a', 'b', 'c', 'd'],
+    oldPoints: {},
+    geometry: null,
+  };
+  const rightPolygon = {
+    modelKey: 'geometryRight',
+    pointNames: ['e', 'f', 'g', 'h', 'i'],
+    oldPoints: {},
+    geometry: null,
+  }
+  const polygons = [
+    leftPolygon,
+    rightPolygon,
+  ];
```

Comment:
I think it might be nicer to create a new class for this, rather than pass around untyped objects, and key names. This would also make it much easier to reuse elsewhere. For example, something like

```ts
class AngleWatcher {
  private oldPoints: Point[];
  private angles: Angle[];
  private exactAngles: number[];

  constructor(private readonly model: Observable, private readonly polygonName: string) {
    // initialise properties...
    model.watch(() => this.update());
  }

  private update() {
    // update properties...
  }

  get angleSum() {
    return total(this.exactAngles);
  }

  get hasOverlap() {
    return this.angleSum > 180 * (this.angles.length - 2);
  }
}

model.set('leftAngleWatcher', new AngleWatcher(model, 'leftPolygon'));
```